lib/SILOptimizer/Utils/ConstExpr.cpp

```diff
@@ -0,0 +1,1288 @@
+  LLVM_DEBUG(llvm::dbgs() << "ConstExpr Opaque Callee: " << *callee << "\n");
+  return evaluator.getUnknown((SILInstruction *)apply, UnknownReason::Default);
+}
+
+// TODO: Refactor this to someplace common, this is defined in Devirtualize.cpp.
+SubstitutionMap getWitnessMethodSubstitutions(SILModule &Module, ApplySite AI,
+                                              SILFunction *F,
+                                              ProtocolConformanceRef CRef);
+
+/// Given a call to a function, determine whether it is a call to a constexpr
+/// function.  If so, collect its arguments as constants, fold it and return
+/// None.  If not, mark the results as Unknown, and return an Unknown with
+/// information about the error.
+llvm::Optional<SymbolicValue>
+ConstExprFunctionState::computeCallResult(ApplyInst *apply) {
+  auto conventions = apply->getSubstCalleeConv();
+
+  // Determine the callee.
+  auto calleeFn = getConstantValue(apply->getOperand(0));
+  if (calleeFn.getKind() != SymbolicValue::Function)
+    return evaluator.getUnknown((SILInstruction *)apply,
+                                UnknownReason::Default);
+
+  SILFunction *callee = calleeFn.getFunctionValue();
+
+  // Verify that we can fold all of the arguments to the call.
+  SmallVector<SymbolicValue, 4> paramConstants;
+  for (unsigned i = 0, e = apply->getNumOperands() - 1; i != e; ++i) {
+    // If any of the arguments is a non-constant value, then we can't fold this
+    // call.
+    auto op = apply->getOperand(i + 1);
+    SymbolicValue argValue = getConstantValue(op);
+    if (!argValue.isConstant())
+      return argValue;
+    paramConstants.push_back(argValue);
+  }
+
+  // If we reached an external function that hasn't been deserialized yet, make
+  // sure to pull it in so we can see its body.  If that fails, then we can't
+  // analyze the function.
+  if (callee->isExternalDeclaration()) {
+    callee->getModule().loadFunction(callee);
+    if (callee->isExternalDeclaration())
+      return computeOpaqueCallResult(apply, callee);
+  }
+
+  // Compute the substitution map for the callee, which maps from all of its
+  // generic requirements to concrete conformances and concrete types.
+  SubstitutionMap calleeSubMap;
+
+  auto calleeFnType = callee->getLoweredFunctionType();
+  if (calleeFnType->getGenericSignature()) {
+    ApplySite AI(apply);
+
+    // Get the substitution map of the call.  This maps from the callee's space
+    // into the caller's world.
+    SubstitutionMap callSubMap;
+    if (calleeFnType->getRepresentation() ==
+        SILFunctionType::Representation::WitnessMethod) {
+      // Get the conformance out of the SILFunctionType and map it into our
+      // current type space.
+      auto conformance = calleeFnType->getWitnessMethodConformance();
+      auto selfTy = conformance.getRequirement()->getSelfInterfaceType();
+      auto conf = substitutionMap.lookupConformance(
```

Comment:
I think you want AI->getSubstitutionMap().lookupConformance()

Reply:
@slavapestov Thanks for pointing this out. You are right that this code is quite broken. @marcrasi I could make a test case (shown below) that crashes it. 
 I think we should use: 
`apply->getSubstitutionMap().subst(substitutionMap).lookupConformance(...)` 
This is because `apply->getSubstitutionMap()` would create a map in which `Self` type of the protocol will be bound to the receiver type of the call site. And, composing it with `substitutionMap` will map it to the actual concrete receiver type tracked by the interpreter. 
Therefore, we would get the right concrete conformance. This seems to fix.

Just for the reference, the test case that bring out this problem is shown below:

```swift
protocol Proto {
  func amethod<U>(_ u: U) -> Int
}

struct S : Proto {
  func amethod<U>(_ u: U) -> Int {
    return 0
  }
}

func callMethod<U, T: Proto>(_ a: T, _ u: U) -> Int {
  return a.amethod(u)
}

func testProtocolMethod() {
  let s = S()
  #assert(callMethod(s, 10) == 0)
}
```